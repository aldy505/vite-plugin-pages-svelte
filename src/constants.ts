export const MODULE_IDS = ['pages-generated', 'virtual:generated-pages', 'virtual:generated-pages-react'];
export const MODULE_ID_VIRTUAL = '/@vite-plugin-pages/generated-pages';
