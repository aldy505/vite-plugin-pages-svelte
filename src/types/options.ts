/**
 * Plugin options.
 */
interface Options {
  /**
   * Relative path to the directory to search for page components.
   * @default 'src/pages'
   */
  pagesDir: string;
  /**
   * Valid file extensions for page components.
   * @default ['svelte']
   */
  extensions: string[];
  /**
   * List of path globs to exclude when resolving pages.
   */
  exclude: string[];
  /**
   * Sync load top level index file
   * @default true
   */
  syncIndex: boolean;
}

export type UserOptions = Partial<Options>;

export interface ResolvedOptions extends Options {
  /**
   * Resolves to the `root` value from Vite config.
   * @default config.root
   */
  root: string;
  /**
   * RegExp to match extensions
   */
  extensionsRE: RegExp;
}

export type ImportMode = 'sync' | 'async';
export type ImportModeResolveFn = (filepath: string) => ImportMode;
