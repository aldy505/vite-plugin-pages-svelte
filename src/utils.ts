import fs from 'fs';
import { resolve, basename } from 'path';
import Debug from 'debug';
import deepEqual from 'deep-equal';
import { ViteDevServer } from 'vite';
import { OutputBundle } from 'rollup';
import { toArray, slash } from '@antfu/utils';
import { ResolvedOptions, Route } from './types';
import { parseSFC, parseCustomBlock } from './parser';
import { MODULE_ID_VIRTUAL } from './constants';

export { toArray, slash };

export const routeBlockCache = new Map<string, Record<string, any>>();

export function extensionsToGlob(extensions: string[]) {
  return extensions.length > 1 ? `{${extensions.join(',')}}` : extensions[0] || '';
}

function isPagesDir(path: string, options: ResolvedOptions) {
  for (const page of options.pagesDir) {
    const dirPath = slash(resolve(options.root, page.dir));
    if (path.startsWith(dirPath)) return true;
  }
  return false;
}

export function isTarget(path: string, options: ResolvedOptions) {
  return isPagesDir(path, options) && options.extensionsRE.test(path);
}

export const debug = {
  hmr: Debug('vite-plugin-pages:hmr'),
  parser: Debug('vite-plugin-pages:parser'),
  gen: Debug('vite-plugin-pages:gen'),
  options: Debug('vite-plugin-pages:options'),
  cache: Debug('vite-plugin-pages:cache'),
  pages: Debug('vite-plugin-pages:pages'),
};

const dynamicRouteRE = /^\[.+\]$/;
export const nuxtDynamicRouteRE = /^_[\s\S]*$/;

export function isDynamicRoute(routePath: string, nuxtStyle = false) {
  return nuxtStyle ? nuxtDynamicRouteRE.test(routePath) : dynamicRouteRE.test(routePath);
}

export function isCatchAllRoute(routePath: string, nuxtStyle = false) {
  return nuxtStyle ? /^_$/.test(routePath) : /^\[\.{3}/.test(routePath);
}

export function resolveImportMode(filepath: string, options: ResolvedOptions) {
  const mode = options.importMode;
  if (typeof mode === 'function') return mode(filepath);

  for (const pageDir of options.pagesDir) {
    if (options.syncIndex && pageDir.baseRoute === '' && filepath === `/${pageDir.dir}/index.vue`) return 'sync';
  }
  return mode;
}

export function pathToName(filepath: string) {
  return filepath.replace(/[_.\-\\/]/g, '_').replace(/[[:\]()]/g, '$');
}

export function findRouteByFilename(routes: Route[], filename: string): Route | null {
  let result = null;
  for (const route of routes) {
    if (filename.endsWith(route.component)) result = route;

    if (!result && route.children) result = findRouteByFilename(route.children, filename);

    if (result) return result;
  }
  return null;
}

export async function getRouteBlock(path: string, options: ResolvedOptions) {
  const content = fs.readFileSync(path, 'utf8');
  const parsed = await parseSFC(content);

  const blockStr = parsed.customBlocks.find((b) => b.type === 'route');
  if (!blockStr) return null;

  const result: Record<string, any> = parseCustomBlock(blockStr, path, options);
  debug.parser('%s: %O', path, result);
  routeBlockCache.set(slash(path), result);

  return result;
}

export function getPagesVirtualModule(server: ViteDevServer) {
  const { moduleGraph } = server;
  const module = moduleGraph.getModuleById(MODULE_ID_VIRTUAL);
  if (module) {
    moduleGraph.invalidateModule(module);
    return module;
  }
  return null;
}

export function replaceSquareBrackets(bundle: OutputBundle) {
  const files = Object.keys(bundle).map((i) => basename(i));
  for (const chunk of Object.values(bundle)) {
    chunk.fileName = chunk.fileName.replace(/(\[|\])/g, '_');
    if (chunk.type === 'chunk') {
      for (const file of files) chunk.code = chunk.code.replace(file, file.replace(/(\[|\])/g, '_'));
    }
  }
}

export async function isRouteBlockChanged(filePath: string, options: ResolvedOptions) {
  debug.cache(routeBlockCache);
  const oldRouteBlock = routeBlockCache.get(filePath);
  const routeBlock = await getRouteBlock(filePath, options);

  debug.hmr('%s old: %O', filePath, oldRouteBlock);
  debug.hmr('%s new: %O', filePath, routeBlock);

  return !deepEqual(oldRouteBlock, routeBlock);
}
