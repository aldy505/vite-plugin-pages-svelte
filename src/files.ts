import fg from 'fast-glob';
import { PageDirOptions, ResolvedOptions } from './types';
import { extensionsToGlob } from './utils';

function getIgnore(exclude: string[]) {
  return ['node_modules', '.git', '**/__*__/**', ...exclude];
}

/**
 * Resolves the page dirs for its for its given globs
 */
export function getPageDirs(pageDirOptions: PageDirOptions, root: string, exclude: string[]): PageDirOptions[] {
  const dirs = fg.sync(pageDirOptions.dir, {
    ignore: getIgnore(exclude),
    onlyDirectories: true,
    dot: true,
    unique: true,
    cwd: root,
  });

  const pageDirs = dirs.map((dir) => ({
    ...pageDirOptions,
    dir,
  }));

  return pageDirs;
}

/**
 * Resolves the files that are valid pages for the given context.
 */
export function getPageFiles(path: string, options: ResolvedOptions): string[] {
  const { exclude, extensions } = options;

  const ext = extensionsToGlob(extensions);

  const files = fg.sync(`**/*.${ext}`, {
    ignore: getIgnore(exclude),
    onlyFiles: true,
    cwd: path,
  });

  return files;
}
