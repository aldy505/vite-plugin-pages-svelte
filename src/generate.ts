import type { PreRoute } from './types/route';
import type { FileOutput } from './types/page';
import { isDynamicRoute, isCatchAllRoute } from './utils/validate';
import { stringifyRoutes } from './stringify';
import { haveChildren } from './crawler/crawler';
import { extname } from 'path';
import { sortRoute } from './utils/route';

/**
 * Generate
 * @param pages
 * @returns
 */
export function generateRoutes(pages: FileOutput[]): PreRoute[] {
  const routes: PreRoute[] = [];

  for (let i = 0; i < pages.length; i++) {
    const node = pages[i].path.split('/')[pages[i].path.split('/').length - 1];
    const fileExt = extname(node);
    const isDynamic = isDynamicRoute(node.replace(fileExt, ''));
    const isCatchAll = isCatchAllRoute(node.replace(fileExt, ''));
    const normalizedName = isDynamic
      ? node
          .replace(fileExt, '')
          .replace(/^\[(\.{3})?/, '')
          .replace(/\]$/, '')
      : node.replace(fileExt, '');
    const normalizedPath = normalizedName.toLowerCase();
    let name: string;
    if (normalizedPath === 'index') {
      name = '/';
    } else {
      if (isCatchAll) {
        name = '/*';
      } else if (isDynamic) {
        name = `/:${normalizedName}`;
      } else {
        name = `/${normalizedPath}`;
      }
    }

    if (!haveChildren(pages[i])) {
      routes.push({
        name,
        path: pages[i].path,
      });
      continue;
    }

    routes.push({
      name,
      children: generateRoutes(pages[i].children as FileOutput[]),
    });
  }

  return routes;
}

/**
 * This pretty much acts as a final codegen for it to be executed by Vite.
 * @param {PreRoute[]} routes
 * @returns {String}
 */
export function generateClientCode(routes: PreRoute[]): string {
  const { imports, stringRoutes } = stringifyRoutes(routes.sort(sortRoute));

  return (
    `${imports.join(';\n')}${imports.length > 1 ? ';' : ''}\n\n` +
    `const routes = ${stringRoutes};\n\n` +
    `export default routes;`
  );
}
