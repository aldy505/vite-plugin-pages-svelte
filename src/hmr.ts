import { ViteDevServer } from 'vite';
import { getPagesVirtualModule, isRouteBlockChanged, isTarget, debug, slash } from './utils';
import { removePage, addPage, updatePage } from './pages';
import { ResolvedOptions, ResolvedPages } from './types';

export function handleHMR(
  server: ViteDevServer,
  pages: ResolvedPages,
  options: ResolvedOptions,
  clearRoutes: () => void,
): void {
  const { ws, watcher } = server;

  function fullReload() {
    // invalidate module
    getPagesVirtualModule(server);
    clearRoutes();
    ws.send({
      type: 'full-reload',
    });
  }

  watcher.on('add', async (file) => {
    const path = slash(file);
    if (isTarget(path, options)) {
      await addPage(pages, path, options);
      debug.hmr('add', path);
      fullReload();
    }
  });
  watcher.on('unlink', (file) => {
    const path = slash(file);
    if (isTarget(path, options)) {
      removePage(pages, path);
      debug.hmr('remove', path);
      fullReload();
    }
  });
  watcher.on('change', async (file) => {
    const path = slash(file);
    if (isTarget(path, options)) {
      const needReload = await isRouteBlockChanged(path, options);
      if (needReload) {
        updatePage(pages, path);
        debug.hmr('change', path);
        fullReload();
      }
    }
  });
}
