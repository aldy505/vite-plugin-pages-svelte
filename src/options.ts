import { resolve } from 'path';
import { UserOptions, ResolvedOptions } from './types';
import { getPageDirs } from './files';
import { toArray, slash } from './utils';

function resolvePageDirs(pagesDir: UserOptions['pagesDir'], root: string, exclude: string[]) {
  pagesDir = toArray(pagesDir);
  return pagesDir.flatMap((pagesDir) => {
    const option = typeof pagesDir === 'string' ? { dir: pagesDir, baseRoute: '' } : pagesDir;

    option.dir = slash(resolve(root, option.dir)).replace(`${root}/`, '');
    option.baseRoute = option.baseRoute.replace(/^\//, '').replace(/\/$/, '');

    return getPageDirs(option, root, exclude);
  });
}

export function resolveOptions(userOptions: UserOptions, viteRoot?: string): ResolvedOptions {
  const {
    pagesDir = ['src/pages'],
    routeBlockLang = 'json5',
    exclude = [],
    syncIndex = true,
    replaceSquareBrackets = false,
    nuxtStyle = false,
    react = false,
    extendRoute,
    onRoutesGenerated,
    onClientGenerated,
  } = userOptions;

  const root = viteRoot || slash(process.cwd());

  const importMode = userOptions.importMode || (react ? 'sync' : 'async');

  const extensions = userOptions.extensions || (react ? ['tsx', 'jsx'] : ['vue', 'ts', 'js']);

  const extensionsRE = new RegExp(`\\.(${extensions.join('|')})$`);

  const resolvedPagesDir = resolvePageDirs(pagesDir, root, exclude);

  const resolvedOptions: ResolvedOptions = {
    pagesDir: resolvedPagesDir,
    routeBlockLang,
    root,
    extensions,
    importMode,
    exclude,
    syncIndex,
    replaceSquareBrackets,
    nuxtStyle,
    react,
    extensionsRE,
    extendRoute,
    onRoutesGenerated,
    onClientGenerated,
  };

  return resolvedOptions;
}
