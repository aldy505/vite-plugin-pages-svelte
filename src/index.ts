import type { Plugin } from 'vite';
import type { ResolvedOptions, UserOptions } from './types/options';
import type { PreRoute } from './types/route';
import type { FileOutput } from './types/page';
import { generateRoutes, generateClientCode } from './generate';
import { debug } from './utils/vite';
import { resolveOptions } from './options';
import { MODULE_IDS, MODULE_ID_VIRTUAL } from './constants';
import { resolvePages } from './pages';
import { handleHMR } from './hmr';

function pagesPlugin(userOptions: UserOptions = {}): Plugin {
  let generatedRoutes: PreRoute[] | null = null;
  let options: ResolvedOptions;
  let pages: FileOutput[];

  return {
    name: 'vite-plugin-pages-svelte',
    enforce: 'pre',
    async configResolved({ root }) {
      options = await resolveOptions(userOptions, root);
      pages = await resolvePages(options);
      debug.options(options);
      debug.pages(pages);
    },
    configureServer(server) {
      handleHMR(server, pages, options, () => {
        generatedRoutes = null;
      });
    },
    resolveId(id) {
      return MODULE_IDS.includes(id) || MODULE_IDS.some((i) => id.startsWith(i)) ? MODULE_ID_VIRTUAL : null;
    },
    load(id) {
      if (id !== MODULE_ID_VIRTUAL) return;

      if (!generatedRoutes) {
        generatedRoutes = generateRoutes(pages);
      }

      const clientCode = generateClientCode(generatedRoutes);
      debug.gen('client code: %O', clientCode);

      return clientCode;
    },
    transform(_code, id) {
      if (!/svelte&type=route/.test(id)) return;
      return {
        code: 'export default {};',
        map: null,
      };
    },
    generateBundle(_options, bundle) {
      bundle;
    },
  };
}

export * from './types/options';
export { generateRoutes };
export default pagesPlugin;
