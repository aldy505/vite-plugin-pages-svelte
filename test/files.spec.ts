import { join } from 'path';
import { slash } from '../src/utils';
import { resolveOptions } from '../src/options';
import { getPageFiles, getPageDirs } from '../src/files';

const options = resolveOptions({});
const testPagesDir = 'test/assets/pages';
const testDeepPagesDir = 'test/assets/deep-pages';

describe('Get files', () => {
  test('Pages file', async () => {
    const files = getPageFiles(testPagesDir, options);
    expect(files.sort()).toMatchSnapshot('page files');
  });
});

describe('Get page dirs', () => {
  test('With glob', async () => {
    const pageDirOptions = {
      dir: slash(join(testDeepPagesDir, '**', 'pages')),
      baseRoute: '',
    };
    const dirs = getPageDirs(pageDirOptions, options.root, options.exclude);
    expect(dirs.sort()).toMatchSnapshot('glob dirs');
  });
});
