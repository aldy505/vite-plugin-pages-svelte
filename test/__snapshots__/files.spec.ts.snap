// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get files Pages file: page files 1`] = `
Array [
  "[...all].vue",
  "[sensor]/current.vue",
  "[userId].vue",
  "about.vue",
  "about/index.vue",
  "blog/[id].vue",
  "blog/index.vue",
  "blog/today/index.vue",
  "components.vue",
  "index.vue",
]
`;

exports[`Get page dirs With glob: glob dirs 1`] = `
Array [
  Object {
    "baseRoute": "",
    "dir": "test/assets/deep-pages/bar/pages",
  },
  Object {
    "baseRoute": "",
    "dir": "test/assets/deep-pages/foo/pages",
  },
]
`;
